internal/eru/agent/agent.go

```diff
@@ -0,0 +1,155 @@
+package agent
+
+import (
+	"context"
+	"strings"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/alphadose/haxmap"
+	"github.com/cockroachdb/errors"
+	"github.com/patrickmn/go-cache"
+	"github.com/projecteru2/core/log"
+	"github.com/projecteru2/yavirt/internal/eru/common"
+	"github.com/projecteru2/yavirt/internal/eru/store"
+	corestore "github.com/projecteru2/yavirt/internal/eru/store/core"
+	storemocks "github.com/projecteru2/yavirt/internal/eru/store/mocks"
+	"github.com/projecteru2/yavirt/internal/eru/types"
+	"github.com/projecteru2/yavirt/internal/service"
+	"github.com/projecteru2/yavirt/internal/utils"
+	"google.golang.org/grpc/status"
+)
+
+type Manager struct {
+	svc    service.Service
+	store  store.Store
+	config *types.Config
+
+	checkWorkloadMutex *sync.Mutex
+	startingWorkloads  *haxmap.Map[string, *utils.RetryTask]
+
+	// storeIdentifier indicates which eru this agent belongs to
+	// it can be used to identify the corresponding core
+	// and all containers that belong to this core
+	storeIdentifier string
+	cas             *utils.GroupCAS
+	wrkStatusCache  *cache.Cache
+
+	mCol *MetricsCollector
+}
+
+func NewManager(
+	ctx context.Context, svc service.Service,
+	config *types.Config, endpoint string,
+	t *testing.T,
+) (*Manager, error) {
+	logger := log.WithFunc("agent.NewManager")
+	interval := time.Duration(2*config.HealthCheck.Interval) * time.Second
+	m := &Manager{
+		config:             config,
+		svc:                svc,
+		cas:                utils.NewGroupCAS(),
+		checkWorkloadMutex: &sync.Mutex{},
+		startingWorkloads:  haxmap.New[string, *utils.RetryTask](),
+		wrkStatusCache:     cache.New(interval, interval),
+	}
+	m.mCol = &MetricsCollector{
+		wrkStatusCache: m.wrkStatusCache,
+	}
+
+	if t == nil {
+		corestore.Init(ctx, config)
+		if m.store = corestore.Get(); m.store == nil {
+			return nil, common.ErrGetStoreFailed
+		}
+	} else {
+		m.store = storemocks.NewFakeStore()
+	}
+	labels := map[string]string{}
+	for _, label := range config.Labels {
+		parts := strings.Split(label, "=")
+		if len(parts) != 2 {
+			return nil, errors.Newf("invalid label %s", label)
+		}
+		labels[parts[0]] = parts[1]
+	}
+	go func() {
+		// Core need to connect to the local grpc server, so sleep 30s here to wait local grpc server up
+		time.Sleep(30 * time.Second)
+		// try to register current node to eru core
+		if _, err := m.store.AddNode(ctx, &types.AddNodeOpts{
+			Nodename: config.Hostname,
+			Endpoint: endpoint,
+			Podname:  config.Podname,
+			Labels:   labels,
+		}); err != nil {
+			e, ok := status.FromError(err)
+			if !ok {
+				logger.Error(ctx, err, "failed to add node")
+				return
+			}
+			if e.Code() == 1031 && strings.Contains(e.Message(), "node already exists") {
```

Comment:
defines a var for (hardcoded) `1031` should be a good idea

Reply:
okay, I'll fix this

Reply:
these code numbers are from core, I just don't want to import too many dependencies, but I just read code and it's ok to import core here.